pkg/registry/parse.go

```diff
@@ -0,0 +1,303 @@
+				if len(decoded.Dependencies) > 0 {
+					df = &decoded
+				}
+			}
+		}
+		if pf == nil {
+			decoded := PropertiesFile{}
+			if err = decodeFileFS(metadata, name, &decoded); err == nil {
+				if len(decoded.Properties) > 0 {
+					pf = &decoded
+				}
+			}
+		}
+	}
+
+	if af != nil {
+		bundle.Annotations = &af.Annotations
+		bundle.Channels = af.GetChannels()
+	} else {
+		return fmt.Errorf("Could not file annotations file")
+	}
+
+	if df != nil {
+		bundle.Dependencies = append(bundle.Dependencies, df.GetDependencies()...)
+	} else {
+		b.log.Info("Could not find optional dependencies file")
+	}
+
+	if pf != nil {
+		bundle.Properties = append(bundle.Properties, pf.Properties...)
+	} else {
+		b.log.Info("Could not find optional properties file")
+	}
+
+	return nil
+}
+
+func (b *bundleParser) derivedProperties(bundle *Bundle) ([]Property, error) {
+	// Add properties from CSV annotations
+	csv, err := bundle.ClusterServiceVersion()
+	if err != nil {
+		return nil, fmt.Errorf("error getting csv: %s", err)
+	}
+	if csv == nil {
+		return nil, fmt.Errorf("bundle missing csv")
+	}
+
+	var derived []Property
+	if len(csv.GetAnnotations()) > 0 {
+		properties, ok := csv.GetAnnotations()[PropertyKey]
+		if ok {
+			if err := json.Unmarshal([]byte(properties), &derived); err != nil {
+				b.log.Warnf("failed to unmarshal csv annotation properties: %s", err)
+			}
+		}
+	}
+
+	if bundle.Annotations != nil && bundle.Annotations.PackageName != "" {
+		pkg := bundle.Annotations.PackageName
+		version, err := bundle.Version()
+		if err != nil {
+			return nil, err
+		}
+
+		value, err := json.Marshal(PackageProperty{
+			PackageName: pkg,
+			Version:     version,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal package property: %s", err)
+		}
+
+		// Annotations file takes precedent over CSV annotations
+		derived = append([]Property{{Type: PackageType, Value: value}}, derived...)
+	}
+
+	providedAPIs, err := bundle.ProvidedAPIs()
+	if err != nil {
+		return nil, fmt.Errorf("error getting provided apis: %s", err)
+	}
+
+	for api := range providedAPIs {
+		value, err := json.Marshal(GVKProperty{
+			Group:   api.Group,
+			Kind:    api.Kind,
+			Version: api.Version,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal gvk property: %s", err)
+		}
+		derived = append(derived, Property{Type: GVKType, Value: value})
+	}
+
+	return joinProperties(derived), nil
+}
+
+// joinProperties deduplicates and returns the given list of properties.
+// It ensures the single package type invariant by retaining the first occurrence of that type in the returned list.
+func joinProperties(properties []Property) []Property {
+	var (
+		foundPkg bool
+		joined   []Property
+		visited  = map[string]struct{}{}
+	)
+	for _, p := range properties {
+		if _, ok := visited[p.String()]; ok {
+			continue
+		}
+		visited[p.String()] = struct{}{}
+
+		switch p.Type {
+		case PackageType:
```

Comment:
is there a way / is it reasonable to not have to worry about this class of problem?

could we ensure properties are sorted by their source (properties.yaml -> csv) and then leave it up to the consumer / resolver to deal with multiple properties (or external validation tools to warn that it's not good to have multiple package properties?)

Reply:
That sounds reasonable.

I'll drop the special case for package properties here, as well as in the expanded properties below.